File: src/Hextech Ordnance/hextechordnance.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
const int BUF = 1<<15;
char inbuf[BUF]; int inpos, inlen;
char next() {
    if (inpos==inlen) {
        inpos=0; inlen=(int)fread(inbuf,1,BUF,stdin);
        //if (!inlen) return EOF;
    }
    return inbuf[inpos++];
}
ll read() {
    char c;
    while (!isdigit(c=next())) {}
    int x=c-48;
    while (isdigit(c=next())) x = x*10+(c-48);
    return x;
}
int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    ll N = read(), S = 0;
    vector<ll> A(N);
    for (int i = 0; i < N; ++i) { A[i] = read(); }
    priority_queue<tuple<ll, ll, ll>, vector<tuple<ll, ll, ll>>, greater<tuple<ll, ll, ll>>> Q;
    unordered_set<int> V;
    for (int i = 0; i < N-1; ++i) {
        ll a = min(A[i], A[i+1]), b = max(A[i], A[i+1]);
        if (b-a <= N) {
            V.insert(i*N+i+1);
            Q.push(make_tuple(b-a, i*N+i+1, a));
        }
    }
    while (!Q.empty() && S++ < N) {
        auto [d, ij, a] = Q.top();
        Q.pop();
        if (d > S) {
            cout << "unstable";
            return 0;
        }
        if (ij > N && V.find(ij-N) == V.end()) {
            V.insert(ij-N);
            ll a2 = min(a, A[ij/N-1]), b2 = max(d+a, A[ij/N-1]);
            Q.push(make_tuple(b2-a2, ij-N, a2));
        }
        if (ij%N < N-1 && V.find(ij+1) == V.end()) {
            V.insert(ij+1);
            ll a2 = min(a, A[ij%N+1]), b2 = max(d+a, A[ij%N+1]);
            Q.push(make_tuple(b2-a2, ij+1, a2));
        }
    }
    cout << (S < N ? "unstable" : "stable");
    return 0;
}

File: src/Hextech Ordnance/hextechordnance.py
import os, io; input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline; from array import *
N = int(input()); A = array('I', map(int, input().split())); S = 0; V = set(); B = [[] for _ in range(N+1)]; d = 0
for i in range(N-1):
    a = min(A[i], A[i+1]); b = max(A[i], A[i+1])
    if b-a <= N: V.add(i*N+i+1); B[b-a].append((i*N+i+1, a))
for S in range(1, N+1):
    while d <= N and not B[d]: d += 1
    if d > S: print('unstable'); exit()
    ij, a = B[d].pop()
    if ij > N and ij-N not in V:
        V.add(ij-N); a2 = min(a, v:=A[ij//N-1]); b2 = max(d+a, v)
        if b2-a2 <= N: B[b2-a2].append((ij-N, a2))
    if ij%N < N-1 and ij+1 not in V:
        V.add(ij+1); a2 = min(a, v:=A[ij%N+1]); b2 = max(d+a, v)
        if b2-a2 <= N: B[b2-a2].append((ij+1, a2))
print('stable')